File: tools/glb.py
#!/usr/bin/env python3
"""`.GLB` 圖庫解包（配 `.GTB` 目錄）。

格式（`docs/formats/02-glb-image-library.md`）：

    GTB = 目錄，每筆 13 bytes
        +0  u8   flag       0 = 調色盤(768 B), 1 = 圖像
        +1  u16  raw_size   解壓後大小（圖像 = w × h，每像素 1 byte）
        +3  u16  comp_size  壓縮後大小
        +5  u32  offset     GLB 內位移
        +9  u16  width
        +11 u16  height

    GLB = 每筆各自是一段獨立的 PKWARE DCL implode 資料（header `00 06`）

調色盤沿用「最近的前一筆調色盤」——GLB 內調色盤與圖像交錯，一個調色盤
管到下一個調色盤出現為止。

用法：
    tools/py.sh tools/glb.py list GRT
    tools/py.sh tools/glb.py extract GRT --out workplace/glb-out
"""

from __future__ import annotations

import argparse
import struct
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))
from dcl import explode  # noqa: E402

ROOT = Path(__file__).resolve().parent.parent
GAME = ROOT / "workplace" / "orig" / "game"

ENTRY = 13


def read_dir(stem: str) -> list[dict]:
    gtb = (GAME / f"{stem}.GTB").read_bytes()
    out = []
    for k in range(len(gtb) // ENTRY):
        e = gtb[k * ENTRY : (k + 1) * ENTRY]
        flag = e[0]
        raw, comp = struct.unpack_from("<HH", e, 1)
        off = struct.unpack_from("<I", e, 5)[0]
        w, h = struct.unpack_from("<HH", e, 9)
        out.append(
            {
                "index": k,
                "flag": flag,
                "kind": "palette" if w == 0 else "image",
                "raw": raw,
                "comp": comp,
                "offset": off,
                "w": w,
                "h": h,
            }
        )
    return out


def load(stem: str) -> tuple[list[dict], bytes]:
    return read_dir(stem), (GAME / f"{stem}.GLB").read_bytes()


def decode_entry(glb: bytes, ent: dict) -> bytes:
    return explode(glb[ent["offset"] : ent["offset"] + ent["comp"]], ent["raw"])


def cmd_list(args) -> int:
    entries, glb = load(args.stem)
    print(f"{args.stem}.GTB: {len(entries)} 筆 / {args.stem}.GLB: {len(glb)} B\n")
    print(f"{'#':>3} {'類型':<8}{'w':>5}{'h':>5}{'壓縮':>8}{'原始':>8}{'位移':>9}  壓縮率")
    for e in entries:
        ratio = e["comp"] / e["raw"] if e["raw"] else 0
        print(
            f"{e['index']:>3} {e['kind']:<8}{e['w']:>5}{e['h']:>5}"
            f"{e['comp']:>8}{e['raw']:>8}{e['offset']:>9}  {ratio:>6.1%}"
        )
    return 0


def cmd_extract(args) -> int:
    from PIL import Image

    entries, glb = load(args.stem)
    out = Path(args.out) / args.stem
    out.mkdir(parents=True, exist_ok=True)

    pal: list[int] | None = None
    pal_from = -1
    n = 0
    for e in entries:
        data = decode_entry(glb, e)
        if e["kind"] == "palette":
            # VGA DAC 是 6-bit（0-63）→ 8-bit 用位元複製擴展 (v<<2)|(v>>4)，
            # 不是 v*255//63。兩者差 1（16→65 vs 64、51→207 vs 206），
            # 逐像素比對原版截圖時這 1 的差會讓整張圖判定為不同。
            pal = [(v << 2) | (v >> 4) for v in data]
            pal_from = e["index"]
            (out / f"{e['index']:02d}-palette.pal").write_bytes(data)
            continue
        if pal is None:
            print(f"  #{e['index']} 沒有可用的調色盤，跳過")
            continue
        img = Image.frombytes("P", (e["w"], e["h"]), data)
        img.putpalette(pal)
        name = f"{e['index']:02d}-{e['w']}x{e['h']}-pal{pal_from:02d}.png"
        img.convert("RGB").save(out / name)
        n += 1
    print(f"{args.stem}: 輸出 {n} 張圖 → {out}")
    return 0


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__)
    sub = ap.add_subparsers(dest="cmd", required=True)
    p = sub.add_parser("list")
    p.add_argument("stem", help="GRT 或 GRT2")
    p = sub.add_parser("extract")
    p.add_argument("stem")
    p.add_argument("--out", default="workplace/glb-out")
    args = ap.parse_args()
    return cmd_list(args) if args.cmd == "list" else cmd_extract(args)


if __name__ == "__main__":
    sys.exit(main())

File: internal/assets/glb.go
package assets

import (
	"encoding/binary"
	"fmt"
)

// .GLB 256 色圖庫 + .GTB 目錄。
//
// GTB entry 固定 13 bytes：
//
//	+0  u8   flag       0 = 調色盤(768 B), 1 = 圖像
//	+1  u16  rawSize    解壓後大小（圖像 = w×h，每像素 1 byte）
//	+3  u16  compSize   壓縮後大小
//	+5  u32  offset     GLB 內位移
//	+9  u16  width
//	+11 u16  height
//
// GLB 的每一筆各自是獨立的 DCL implode 資料（header 00 06）。

// GTBEntrySize 是目錄每筆的固定大小。
const GTBEntrySize = 13

// GTBEntry 是圖庫目錄的一筆。
type GTBEntry struct {
	Index    int
	Flag     byte
	RawSize  int
	CompSize int
	Offset   int
	W, H     int
}

// IsPalette 回報這筆是不是調色盤。調色盤的 W/H 是 0、RawSize 是 768。
func (e GTBEntry) IsPalette() bool { return e.W == 0 || e.H == 0 }

// ParseGTB 解析 .GTB 目錄，並檢查自洽性。
func ParseGTB(data []byte) ([]GTBEntry, error) {
	if len(data)%GTBEntrySize != 0 {
		return nil, fmt.Errorf("assets: .GTB 大小 %d 不是 %d 的倍數", len(data), GTBEntrySize)
	}
	n := len(data) / GTBEntrySize
	out := make([]GTBEntry, 0, n)
	for k := 0; k < n; k++ {
		b := data[k*GTBEntrySize:]
		e := GTBEntry{
			Index:    k,
			Flag:     b[0],
			RawSize:  int(binary.LittleEndian.Uint16(b[1:])),
			CompSize: int(binary.LittleEndian.Uint16(b[3:])),
			Offset:   int(binary.LittleEndian.Uint32(b[5:])),
			W:        int(binary.LittleEndian.Uint16(b[9:])),
			H:        int(binary.LittleEndian.Uint16(b[11:])),
		}
		// 自洽性：圖像的 rawSize 必須等於 w×h，調色盤必須是 768。
		want := e.W * e.H
		if e.IsPalette() {
			want = 768
		}
		if e.RawSize != want {
			return nil, fmt.Errorf("assets: .GTB #%d rawSize=%d，預期 %d (%dx%d)",
				k, e.RawSize, want, e.W, e.H)
		}
		out = append(out, e)
	}
	return out, nil
}

// GLB 是一組目錄加上壓縮資料。
type GLB struct {
	Entries []GTBEntry
	Data    []byte
}

// OpenGLB 綁定 .GTB 與 .GLB，並驗證兩者對得起來。
func OpenGLB(gtb, glb []byte) (*GLB, error) {
	entries, err := ParseGTB(gtb)
	if err != nil {
		return nil, err
	}
	for i, e := range entries {
		if e.Offset+e.CompSize > len(glb) {
			return nil, fmt.Errorf("assets: .GTB #%d 指到 %d..%d，超出 .GLB 的 %d bytes",
				i, e.Offset, e.Offset+e.CompSize, len(glb))
		}
		// 每筆各自是完整的 DCL stream，起點必須是合法 header。
		h := glb[e.Offset : e.Offset+2]
		if h[0] > 1 || h[1] < 4 || h[1] > 6 {
			return nil, fmt.Errorf("assets: .GTB #%d 起點不是 DCL header (%02x %02x)", i, h[0], h[1])
		}
		if i > 0 {
			prev := entries[i-1]
			if prev.Offset+prev.CompSize != e.Offset {
				return nil, fmt.Errorf("assets: .GTB #%d 的位移 %d 接不上前一筆的結尾 %d",
					i, e.Offset, prev.Offset+prev.CompSize)
			}
		}
	}
	return &GLB{Entries: entries, Data: glb}, nil
}

// Raw 解出第 i 筆的原始資料。
func (g *GLB) Raw(i int) ([]byte, error) {
	if i < 0 || i >= len(g.Entries) {
		return nil, fmt.Errorf("assets: .GLB 索引 %d 超出範圍 0..%d", i, len(g.Entries)-1)
	}
	e := g.Entries[i]
	return Explode(g.Data[e.Offset:e.Offset+e.CompSize], e.RawSize)
}

// Palette 解出第 i 筆的調色盤。
func (g *GLB) Palette(i int) (Palette, error) {
	if !g.Entries[i].IsPalette() {
		return nil, fmt.Errorf("assets: .GLB #%d 不是調色盤", i)
	}
	raw, err := g.Raw(i)
	if err != nil {
		return nil, err
	}
	return ParseVGAPalette(raw)
}

// PaletteFor 回傳管轄第 i 筆圖像的調色盤索引。
//
// 調色盤與圖像在 .GLB 內交錯排列，一個調色盤管到下一個調色盤出現為止，
// 所以要往前找最近的一個。找不到回 -1。
func (g *GLB) PaletteFor(i int) int {
	for k := i; k >= 0; k-- {
		if g.Entries[k].IsPalette() {
			return k
		}
	}
	return -1
}

// Sprite 是 .GLB 裡的一張 256 色圖。
//
// 索引 0 是透明色：疊圖時不寫入，露出底圖。書法標題就是這樣疊上去的。
type Sprite struct {
	W, H int
	Pix  []byte // 每像素 1 byte，值 0-255
}

// TransparentIndex 是 .GLB 圖像的透明色索引。
const TransparentIndex = 0

// Sprite 解出第 i 筆圖像。
func (g *GLB) Sprite(i int) (*Sprite, error) {
	e := g.Entries[i]
	if e.IsPalette() {
		return nil, fmt.Errorf("assets: .GLB #%d 是調色盤不是圖像", i)
	}
	raw, err := g.Raw(i)
	if err != nil {
		return nil, err
	}
	return &Sprite{W: e.W, H: e.H, Pix: raw}, nil
}
